<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<title>Cufflinks :: Center for Bioinformatics and Computational Biology</title>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8" />
<link rel="stylesheet" type="text/css" href="css/style.css" media="screen" />
</head>
<body>
<div id="wrap">
  <div id="top">
    <div class="lefts">
    <table width="100%" cellpadding="2">
      <tr><td>
        <a href="./index.html"><h1>Cufflinks</h1></a>
        <h2>Transcript assembly, differential expression, and differential regulation for RNA-Seq</h2>
      </td><td align="right" valign="middle">
        <a href="http://www.mcb.berkeley.edu/"><img style="vertical-align:middle;padding-top:4px" border=0 src="images/UCBerkeley-seal.scaled.gif"></img></a>&nbsp;
        <a href="http://www.cbcb.umd.edu/"><img style="vertical-align:middle;padding-top:4px" border=0 src="images/cbcb_logo.gif"></a>&nbsp;&nbsp;
      </td></tr>
    </table>
    </div>
  </div>
  <div id="subheader">
  <table width="100%"><tr>
  <td>
  		<strong>Please Note</strong> If you have questions 
		  about how to use Cufflinks or would like more information about 
		  the software, please email <a href="mailto:tophat.cufflinks@gmail.com">the Cufflinks team</a>, 
		  though we ask you to have a look at the <a href="http://dx.doi.org/10.1038/nbt.1621">paper</a> 
		  and the  <a href="http://www.nature.com/nbt/journal/v28/n5/extref/nbt.1621-S1.pdf">
		  supplemental methods</a> first, as your question may be answered there.
	</td><td align=right valign=middle>
  </td></tr>
  </table>
  </div>
  <div id="main">
    <div id="rightside">
		<h2>Site Map</h2>
		  <div class="box">
		    <ul>
		      <li><a href="index.html">Home</a></li>
		      <li><a href="tutorial.html">Getting started</a></li>
		      <li><a href="manual.html">Manual</a></li>
			  <li><a href="howitworks.html">How Cufflinks works</a></li>
		      <li><a href="faq.html">FAQ</a></li>
		    </ul>
		  </div>
		
		  <h2><u>News and updates</u></h2>
		  <div class="box">
		    <ul>
		      <table width="100%">
				<tr><td>New releases and related tools will be announced through the <a href="https://lists.sourceforge.net/lists/listinfo/bowtie-bio-announce"><b>mailing list</b></a></td></tr>
			  </table>
		    </ul>
		  </div>
		
		  <h2><u>Getting Help</u></h2>
		  <div class="box">
		    <ul>
		      <table width="100%">
				<tr><td>Questions about Cufflinks should be sent to <a href="mailto:tophat.cufflinks@gmail.com"><b>tophat.cufflinks@gmail.com</b></a>.  Please do not email technical questions to Cufflinks contributors directly.</td></tr>
			  </table>
		    </ul>
		  </div>
		
		  <a href="./downloads"><h2><u>Releases</u></h2></a>
		  <div class="box">
		    <ul>
		      <table width="100%">
				<tr><td>version 0.9.3 (BETA)</td> <td align="right">11/30/2010</td></tr>
				<tr>
		      		<td><a href="./downloads/cufflinks-0.9.3.tar.gz" onClick="javascript: pageTracker._trackPageview('/downloads/cufflinks_source'); ">&nbsp;&nbsp;&nbsp;Source code</a></td>
		      	</tr>
				<tr>
					<td><a href="./downloads/cufflinks-0.9.3.Linux_x86_64.tar.gz" onClick="javascript: pageTracker._trackPageview('/downloads/cufflinks'); ">&nbsp;&nbsp;&nbsp;Linux x86_64 binary</a></td>
				</tr>
				<tr>
					<td><a href="./downloads/cufflinks-0.9.3.OSX_x86_64.tar.gz" onClick="javascript: pageTracker._trackPageview('/downloads/cufflinks'); ">&nbsp;&nbsp;&nbsp;Mac OS X x86_64 binary</a></td>
				</tr>
			  </table>
		    </ul>
		  </div>
		
		  <h2>Related Tools</h2>
		  <div class="box">
		    <ul>
		      <li><a href="http://tophat.cbcb.umd.edu/">TopHat</a>: Alignment of short RNA-Seq reads</li>
		      <li><a href="http://bowtie.cbcb.umd.edu">Bowtie</a>: Ultrafast short read alignment</li>
		    </ul>
		  </div>
		  <h2>Publications</h2>
		  <div class="box">
		    <ul>
			  <li style="font-size: x-small; line-height: 130%"><p>Trapnell C, Williams BA, Pertea G, Mortazavi AM, Kwan G, van Baren MJ, Salzberg SL, Wold B, Pachter L.<b> <a href="http://dx.doi.org/10.1038/nbt.1621">Transcript assembly and quantification by RNA-Seq reveals unannotated transcripts and isoform switching during cell differentiation</b></a> <br/><i><a href="http://www.nature.com/nbt">Nature Biotechnology</a></i> doi:10.1038/nbt.1621</p><br/></li>
 			  <li style="font-size: x-small; line-height: 130%"><p>Roberts A, Trapnell C, Donaghey J, Rinn JL, Pachter L.<b> <a href="http://genomebiology.com/2011/12/3/R22/abstract">Improving RNA-Seq expression estimates by correcting for fragment bias</b></a> <br/><i><a href="http://www.genomebiology.com">Genome Biology</a></i> doi:10.1186/gb-2011-12-3-r22</p><br/></li>
		    </ul>
		  </div>

		  <h2>Contributors</h2>
		  <div class="box">
		    <ul>
		      <li><a href="http://www.cs.umd.edu/~cole/">Cole Trapnell</a></li>
		      <li><a href="http://www.cs.berkeley.edu/~adarob/">Adam Roberts</a></li>
			  <li>Geo Pertea</li>
			  <li>Brian Williams</li>
			  <li><a href="http://wormlab.caltech.edu/members/">Ali Mortazavi</a></li>
			  <li>Gordon Kwan</a></li>
			  <li>Jeltje van Baren</a></li>
		      <li><a href="http://www.cbcb.umd.edu/~salzberg/">Steven Salzberg</a></li>
		      <li><a href="http://biology.caltech.edu/Members/Wold">Barbara Wold</a></li>
			  <li><a href="http://www.math.berkeley.edu/~lpachter/">Lior Pachter</a></li>
		    </ul>
		  </div>

		  <h2>Links</h2>
		  <div class="box">
		    <ul>
		      <li><a href="http://bio.math.berkeley.edu/">Berkeley LMCB</a></li>
		      <li><a href="http://www.cbcb.umd.edu/">UMD CBCB</a></li>
			  <li><a href="http://woldlab.caltech.edu/">Wold Lab</a></li>
		    </ul>
		  </div>          
          
    </div> <!-- End of "rightside" -->
    <div id="leftside">
  	  <table><tr><td cellpadding=7>
  	  <h1>Frequently Asked Questions</h1><br/>
      <div id="toc">
  	    <ul>
 	    <li><a href="#hass">What's the difference between FPKM and RPKM?</a></li>
 	    <li><a href="#hssd">Can Cufflinks handle strand-specific RNA-Seq data?</li>
 	  	<li><a href="#hdis">Can I use Cufflinks with RNA-Seq data from Bacteria?</a></li>
 	  	<li><a href="#hsbi">I want to find differentially expressed genes.  Can I use Cufflinks in conjunction with count-based differential expression packages?</a></li>
		<li><a href="#hsbi">Will you please report how many fragments come from each transcript in a future release?</a></li>
		<li><a href="#hsbi">How is Cuffdiff different from other differential expression tools?</a></li>
 	  	<li><a href="#hmul">Can I use Cufflinks to find SNPs or RNA editing sites?</a></li>
  	  	<li><a href="#hrga">Does Cufflinks discover gene fusions or trans-spliced transcripts?</a></li>
 	  	<li><a href="#hbam">Does Cufflinks support both BAM and SAM?</a></li>
	   	<li><a href="#huqn">What does "NOTEST" mean in Cuffdiff's output?</a></li>
	    <li><a href="#reps">Why does Cufflinks occasionally report "FAIL" for some genes?</a></li>
	 	<li><a href="#reps">How many replicates do I need?  Do I need technical replicates as well as biological ones?</a></li>
		<li><a href="#reps">How does Cufflinks calculate its p-values?</a></li>
		<li><a href="#reps">I'm looking for a GTF file with annotated transcripts.  Do you provide annotation files?</a></li>
		<li><a href="#h99p">I'm trying to assemble a sample.  Cufflinks is almost done, but it seems to be hanging at "99% complete".  What's going on?</a></li>
		<li><a href="#reps">Can I use Cuffdiff on time-series data?</a></li>
		<li><a href="#reps">How can I visualize by Cufflinks and Cuffdiff results?</a></li>
 	    <li><a href="#reps">I'm having trouble installing Cufflinks on my system.  Can you help me?</a></li>
		<li><a href="#reps">I'm getting a lot of single-exon, low FPKM transfrags in my asssembly.  What are these?</a></li>
		

  	    </ul><br/>
  	  </div>
  	  <h2 id="hssd">What's the difference between FPKM and RPKM?</h2><br/><p> 
	  They're almost the same thing.  RPKM stands for Reads Per Kilobase of transcript per Million mapped reads.  FPKM stands for Fragments Per Kilobase of transcript per Million mapped reads.  In RNA-Seq, the relative expression of a transcript is proportional to the number of cDNA fragments that originate from it. Paired-end RNA-Seq experiments produce two reads per fragment, but that doesn't necessarily mean that both reads will be mappable.  For example, the second read is of poor quality. If we were to count reads rather than fragments, we might double-count some fragments but not others, leading to a skewed expression value. Thus, FPKM is calculated by counting fragments, not reads.   
  	  </p><br/>

	  <h2 id="hssd">Can Cufflinks handle strand-specific RNA-Seq data?</h2><br/><p> 
	  Yes.  If you mapped your reads with TopHat using the strand-specific mode or manually added
  	  XS tags, Cufflinks will automatically treat your data as strand-specific.  If this is not the case,
  	  you should use the <tt>--library-type</tt> option along with the type of strand-specific protocol that
  	  was used to generate your data.  More information can be found <a href=manual.html#library>here</a>.
  	  </p><br/>

	  <h2 id="hssd">Can I use Cufflinks with RNA-Seq data from Bacteria?</h2><br/><p> 
	  Sure, with one or two caveats.  We don't recommend assembling bacterial transcripts using Cufflinks at first. If you're working on a new bacterial genome, consider using a computational gene finding application such as <a href="http://www.cbcb.umd.edu/software/glimmer/">Glimmer</a>.  These tools are very mature and work well.  With an annotation, you can use Cufflinks and Cuffdiff to profile expression and look for differences.  Also, not that Cufflinks doesn't handle circular genomes in any special way, so make sure your genome and annotation are suitably linearized.</p><br/>
	
	  <h2 id="hssd">I want to find differentially expressed genes.  Can I use Cufflinks in conjunction with count-based differential expression packages?</h2><br/><p>
	  It's possible, but we strongly advise against this.  Current count-based differential expression tools are poorly suited to differential expression analysis in genomes with alternatively spliced genes.  The main reason for this is that when a gene has multiple isoforms, a change in the total number of reads or fragments from that gene doesn't correspond to a change in expression for that gene. 
	  </p><br/>
	
	  <h2 id="hbam">Does Cufflinks support both BAM and SAM?</h2><br/><p> 
	  Yes. If a SAM is supplied, an error message will be output that the file is not a valid BAM file.
	  However, Cufflinks will recognize this and treat the file as a SAM.  When using a SAM file, you should
	  include a proper header or ensure that the reads are lexicographically by chromosome and then numerically
	  by left position.  You can accomplish this sorting with the command <tt>sort -k3,3 -k4,4n in.sam > out.sam</tt>.
	  </p><br/>
	
	  <h2 id="h99p">I'm trying to assemble a sample.  Cufflinks is almost done, but it seems to be hanging at "99% complete".  What's going on?</h2><br/><p> 
	  Cufflinks spawns threads for each locus to assemble and quantitate the "bundle" of reads in that locus.  Some loci may have more reads and more complicated alternative
	  splicing than others, which requires more CPU cycles.  These bundles can continue processing long after all others have completed, leading
	  to this behavior.  You may be able to decrease the number of such bundles by masking out ribosomal and mitochondrial RNA using the <tt>-M/--mask-file</tt> option
	  described in the <a href="manual.html">Manual</a>.
	  </p><br/>
	
	</div>
  </div>
  <div id="footer">
  	<table width="100%" cellspacing=15><tr><td>
    This research was supported in part by NIH grants R01-LM06845 and R01-GM083873, NSF grant CCF-0347992 and the Miller Institute for Basic
		Research in Science at UC Berkeley.
    </td><td align=right>
    Administrator: <a href="mailto:cole@cs.umd.edu">Cole Trapnell</a>. Design by <a href="http://www.free-css-templates.com" title="Design by David Herreman">David Herreman</a>
    </td></tr></table>
  </div>
</div>

<!-- Google analytics code -->
<script type="text/javascript">
var gaJsHost = (("https:" == document.location.protocol) ? "https://ssl." : "http://www.");
document.write(unescape("%3Cscript src='" + gaJsHost + "google-analytics.com/ga.js' type='text/javascript'%3E%3C/script%3E"));
</script>
<script type="text/javascript">
try {
var pageTracker = _gat._getTracker("UA-6101038-2");
pageTracker._trackPageview();
} catch(err) {}</script>
<!-- End Google analytics code -->
</body>
</html>
