<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<title>Cufflinks :: Center for Bioinformatics and Computational Biology</title>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8" />
<link rel="stylesheet" type="text/css" href="css/style.css" media="screen" />
</head>
<body>
<div id="wrap">
  <div id="top">
    <div class="lefts">
    <table width="100%" cellpadding="2">
      <tr><td>
        <a href="./index.html"><h1>Cufflinks</h1></a>
        <h2>Transcript assembly, differential expression, and differential regulation for RNA-Seq</h2>
      </td><td align="right" valign="middle">
        <h1><a href="http://www.cbcb.umd.edu/"><img border=0 src="images/cbcblogo.gif"></img></a>&nbsp;&nbsp;</h1>
      </td></tr>
    </table>
    </div>
  </div>
  <div id="subheader">
  <table width="100%"><tr>
  <td>
  <strong>Cufflinks</strong> assembles transcripts, estimates their abundances, 
    and tests for <b>differential expression</b> and <b>regulation</b> in RNA-Seq samples.  
    It accepts aligned RNA-Seq reads 
	and assembles the alignments into a parsimonious set of transcripts.
	Cufflinks then estimates the relative abundances of these transcripts based
	on how many reads support each one.
	 	<br/> <br/>
	Cufflinks is a collaborative effort between the <a href="http://bio.math.berkeley.edu/">
	Laboratory for Mathematical and Computational Biology</a>, led by Lior Pachter at UC Berkeley, 
	Steven Salzberg's group at the University of Maryland 
	<a href="http://cbcb.umd.edu">Center for Bioinformatics and Computational 
	Biology</a>, and <a href="http://woldlab.caltech.edu/">Barbara Wold's lab</a> at Caltech.
	<br/> <br/>
	Cufflinks is provided under the OSI-approved <a href="http://en.wikipedia.org/wiki/Boost_Software_License">Boost License</a>
	</td><td align=right valign=middle>
  <a href="http://opensource.org"><img border=0 alt="Open Source Software" src="images/osi-certified.gif"></img></a>
  </td></tr>
  </table>
  </div>
  <div id="main">
    <div id="rightside">
		<h2>Site Map</h2>
		  <div class="box">
		    <ul>
		      <li><a href="index.html">Home</a></li>
		      <li><a href="tutorial.html">Getting started</a></li>
		      <li><a href="manual.html">Manual</a></li>
			  <li><a href="howitworks.html">How Cufflinks works</a></li>
		    </ul>
		  </div>
		
		  <h2><u>News and updates</u></h2>
		  <div class="box">
		    <ul>
		      <table width="100%">
				<tr><td>New releases and related tools will be announced through the <a href="https://lists.sourceforge.net/lists/listinfo/bowtie-bio-announce"><b>mailing list</b></a></td></tr>
			  </table>
		    </ul>
		  </div>
		
		  <a href="./downloads"><h2><u>Releases</u></h2></a>
		  <div class="box">
		    <ul>
		      <table width="100%">
				<tr><td>version 0.9.0 (BETA)</td> <td align="right">9/XXX/2010</td></tr>
				<tr>
		      		<td><a href="./downloads/cufflinks-0.9.0.tar.gz" onClick="javascript: pageTracker._trackPageview('/downloads/cufflinks_source'); ">&nbsp;&nbsp;&nbsp;Source code</a></td>
		      	</tr>
				<tr>
					<td><a href="./downloads/cufflinks-0.9.0.Linux_x86_64.tar.gz" onClick="javascript: pageTracker._trackPageview('/downloads/cufflinks'); ">&nbsp;&nbsp;&nbsp;Linux x86_64 binary</a></td>
				</tr>
				<tr>
					<td><a href="./downloads/cufflinks-0.9.0.OSX_x86_64.tar.gz" onClick="javascript: pageTracker._trackPageview('/downloads/cufflinks'); ">&nbsp;&nbsp;&nbsp;Mac OS X x86_64 binary</a></td>
				</tr>
			  </table>
		    </ul>
		  </div>
		
		  <h2>Related Tools</h2>
		  <div class="box">
		    <ul>
		      <li><a href="http://tophat.cbcb.umd.edu/">TopHat</a>: Alignment of short RNA-Seq reads</li>
		      <li><a href="http://bowtie.cbcb.umd.edu">Bowtie</a>: Ultrafast short read alignment</li>
		    </ul>
		  </div>
		
		  <h2>Publications</h2>
		  <div class="box">
		    <ul>
			  <li style="font-size: x-small; line-height: 130%"><p>Trapnell C, Williams BA, Pertea G, Mortazavi AM, Kwan G, van Baren MJ, Salzberg SL, Wold B, Pachter L.<b> <a href="http://dx.doi.org/10.1038/nbt.1621">Transcript assembly and quantification by RNA-Seq reveals unannotated transcripts and isoform switching during cell differentiation</b></a> <br/><i><a href="http://www.nature.com/nbt">Nature Biotechnology</a></i> doi:10.1038/nbt.1621</p><br/></li>
			  <li style="font-size: x-small; line-height: 130%"><p>Trapnell C, Pachter L, Salzberg SL. <a href="http://bioinformatics.oxfordjournals.org/cgi/content/abstract/btp120"><b>TopHat: discovering splice junctions with RNA-Seq</b></a>. <i><a href="http://bioinformatics.oxfordjournals.org">Bioinformatics</a></i>  doi:10.1093/bioinformatics/btp120</p><br/></li>
		      <li style="font-size: x-small; line-height: 130%"><p>Langmead B, Trapnell C, Pop M, Salzberg SL. <a href="http://genomebiology.com/2009/10/3/R25"><b>Ultrafast and memory-efficient alignment of short DNA sequences to the human genome</b></a>. <br/><i><a href="http://genomebiology.com">Genome Biology</a></i> <b>10</b>:R25.</li>
		    </ul>
		  </div>

		  <h2>Contributors</h2>
		  <div class="box">
		    <ul>
		      <li><a href="http://www.cs.umd.edu/~cole/">Cole Trapnell</a></li>
   		      <li><a href="http://www.cs.berkeley.edu/~adarob/">Adam Roberts</a></li>
			  <li>Geo Pertea</li>
			  <li>Brian Williams</li>
			  <li><a href="http://wormlab.caltech.edu/members/">Ali Mortazavi</a></li>
			  <li>Gordon Kwan</a></li>
			  <li>Jeltje van Baren</a></li>
		      <li><a href="http://www.cbcb.umd.edu/~salzberg/">Steven Salzberg</a></li>
		      <li><a href="http://biology.caltech.edu/Members/Wold">Barbara Wold</a></li>
			  <li><a href="http://www.math.berkeley.edu/~lpachter/">Lior Pachter</a></li>
		    </ul>
		  </div>

		  <h2>Links</h2>
		  <div class="box">
		    <ul>
		      <li><a href="http://bio.math.berkeley.edu/">Berkeley LMCB</a></li>
		      <li><a href="http://www.cbcb.umd.edu/">UMD CBCB</a></li>
			  <li><a href="http://woldlab.caltech.edu/">Wold Lab</a></li>
		    </ul>
		  </div>          

    </div> <!-- End of "rightside" -->
    <div id="leftside">
	
		<h2>0.9.0 release - 9/XXX/2010</h2>
	      <p>This release includes significant bug fixes as well as some major new features.  Enhancements include:
		<ul>
			<li>Thanks to the work of Adam Roberts, Cufflinks now automatically learns certain properties of each RNA-Seq 
				library you analyze with it, including sequence specific biases.  Transcript and gene abundance estimates
				are significantly more accurate when run with our (optional) new bias correction technology.  
				A manuscript describing these changes is in preparation. </li>
			<li>Due to Adam's changes, we do <b>not</b> recommend that you supply the mean and standard deviations 
				of the fragment length distribution for your library.  Cufflinks now learns these automatically.
			<li>Cuffdiff now allows you to supply reads from multiple technical or biological replicates, 
				and adjusts transcript and gene abundance estimates and their confidence intervals accordingly.</li>
			<li>Cuffdiff will by default report testing results between all pairs of samples, rather than treating 
				samples as a time series.  Time series analysis is still supported via the <tt>--time-series</tt> option.</li>
			<li>Cufflinks and Cuffdiff now offer improved compatibility with reads aligned with Life Technology's BioScope aligner.  
				Strand specificity SOLiD RNA-Seq runs are fully supported.</li>
			<li>Cufflinks and Cuffdiff support quantile normalization of FPKM.</li>
			<li>Cufflinks and Cuffdiff now support both BAM and SAM files.  Both file types require that the chromosome names in
				 header <tt>@SQ</tt> records, if present, <b>must</b> be sorted in the same order as the reads.</li>
			<li>Cufflinks and Cuffdiff are much less verbose in their output by default, but each supports a "verbose" 
				mode with a text progress bar for users that want to keep an eye on their runs</li>
			<li>In response to user requests, FPKM values are now provided by Cufflinks and Cuffdiff, 
				even if their expression levels are zero, in an effort to make downstream analysis of Cuffdiff easier.</li>
			<li>There are a number of new command line options that users may wish to set under certain circumstances.</li>
		</ul>
		<p>Note that building Cufflinks from source now requires the <a href="http://samtools.sourceforge.net">SAM tools</a>.  
		See the installation instructions for more details. </p>
		  <br/>
	
	  <h2>0.8.3 build update - 7/2/2010</h2>
      <p> The 0.8.3 release on 6/30 included a broken version of Cuffdiff.  This and a few other minor issues have been resolved and the builds and source have been updated.
	  </p>
	  <br/>
	  <h2>0.8.3 release - 6/30/2010</h2>
      <p> Fixes and accuracy improvements to the Cufflinks assembler.  This update is strongly recommended for users interested in identifying novel transcripts. Other changes:
	    <ul>
	      <li>Additional expression reporting fixes</li>
		  <li>Output from Cufflinks and Cuffdiff can now be directed to a separate directory.</li>
	    </ul>
	  </p>
	  <br/>
	
	  <h2>Cufflinks paper published - 5/2/2010</h2>
      <p> Our <a href="http://dx.doi.org/10.1038/nbt.1621">paper</a> on Cufflinks has appeared 
	  at <i><a href="http://www.nature.com/nbt">Nature Biotechnology</a></i>.  In a 
	  study of developing mouse muscle cells, we identified thousands of new 
	  transcripts and detected differential splicing and promoter use in 
	  hundreds of genes by examining changes in isoform abundance levels.  
	  The paper is aimed at a general audience, so much of the algorithmic 
	  and mathematical detail can be found in the <a href="http://www.nature.com/nbt/journal/vaop/ncurrent/extref/nbt.1621-S1.pdf">supplemental methods</a> (WARNING: 2 MB PDF). 
  	  </p>
	  <br/>
	  
	
	  <h2>0.8.2 release - 3/26/2010</h2>
      <p> Numerous bug fixes and significant performance improvements in 
	  Cufflinks and Cuffdiff.  Note that some the <strong>formats</strong> of some files <strong>have changed</strong>. Other changes:
	    <ul>
	      <li>Expression values were not being reported for a small number of genes
		      when using certain annotation files.</li>
		  <li>Fixed a divide-by-zero error in Cuffdiff</li>
	      <li>Various GTF parser fixes.</li>
		  <li><tt>.expr</tt> files produced by Cufflinks now contain FPKM confidence interval columns</li>
		  <li>Isoform filtering was being innappropriately applied when working with annotation in Cufflinks (via <tt>--GTF</tt>)</li>
	      <li>The <tt>.tmap</tt> files reported by Cuffdiff contained zero in FPKM field.</li>
		  <li>New command line options <tt>--num-importance-samples</tt> and <tt>--max-mle-iterations</tt> allow advanced users to influence of the transcript abundance estimation.
	    </ul>
	  </p>
	  <br/>
	
	  <h2>0.8.1 release - 2/13/2010</h2>
      <p> This is a minor fix release.  Cufflinks 0.8.0 had some lingering 
	  references to "RPKM", which have been replaced with "FPKM".  Reported 
	  expression values have always been in FPKM, this is simply a nomenclature 
	  change to better reflect what RNA-Seq actually measures.  See the section on FPKM in 
	  the description of how Cufflinks <a href="howitworks.html">works</a>
	  </p>
	  <br/>
	
      <h2>0.8.0 release - 2/5/2010</h2>
      <p> We are happy to announce a major update to Cufflinks that introduces 
	  some powerful new features and includes a number of performance improvement
	  and bug fixes.  Highlights include:
	
	  <p>
	  <ul>
	  <li>Cufflinks now includes a new tool, "Cuffdiff", which performs testing
		for differential expression, splicing, promoter use, and coding sequence
		output on two or more RNA-Seq samples.  See the greatly expanded 
		<a href="manual.html">manual</a> for details.</li>
		
	  <li>Cuffcompare now reports a file containing the "union" of all transfrags
		in the files you give it as input, greatly simplifying downstream validatation
		of novel transcripts.</li>
		
	  <li>Cufflinks' assembler has been overhauled and optimized, resulting in
		a speedup of 4-5 times over version 0.7.0, and a <b>greatly</b> reduced
		memory footprint.  Phasing of splicing events has also been improved.</li>
	
	  <li>Many bugfixes.</li>
	  </ul>
	  <br/>
      <h2>0.7.0 release - 9/26/2009</h2>
      <p> The first public release of Cufflinks is now available for download.
	   Cufflinks is a program for the comparative assembly of transcripts and 
	   the estimation of their abundances in an RNA-Seq experiment.  It runs on 
	   Linux and OS X.  It also comes with a tool to track transcripts across
	   multiple samples, for example in a time course of RNA-Seq.
	
	  <p>Cufflinks takes as input a file of alignments in <a href="http://samtools.sourceforge.net">SAM</a> 
		format, and reports transfrags in 
		<a href="http://mblab.wustl.edu/GTF22.html">GTF</a> format.  You can 
		use <a href="http://tophat.cbcb.umd.edu">TopHat</a> to align your 
		reads, as TopHat reports its alignments in SAM format.
		
   	  <p>This software is a work in progress - it is a beta release, and new 
	  features will continue to be added over the next couple of weeks.  To 
	suggest a feature or report a bug, please email <a href="mailto:cole@cs.umd.edu">Cole Trapnell</a>
	</div>
  </div>
  <div id="footer">
  	<table width="100%" cellspacing=15><tr><td>
    	This research was supported in part by NIH grants R01-LM06845 and R01-GM083873, NSF grant CCF-0347992 and the Miller Institute for Basic
			Research in Science at UC Berkeley.
    </td><td align=right>
    Administrator: <a href="mailto:cole@cs.umd.edu">Cole Trapnell</a>. Design by <a href="http://www.free-css-templates.com" title="Design by David Herreman">David Herreman</a>
    </td></tr></table>
  </div>
</div>

<!-- Google analytics code -->
<script type="text/javascript">
var gaJsHost = (("https:" == document.location.protocol) ? "https://ssl." : "http://www.");
document.write(unescape("%3Cscript src='" + gaJsHost + "google-analytics.com/ga.js' type='text/javascript'%3E%3C/script%3E"));
</script>
<script type="text/javascript">
try {
var pageTracker = _gat._getTracker("UA-6101038-2");
pageTracker._trackPageview();
} catch(err) {}</script>
</script>
<!-- End google analytics code -->

</body>
</html>
