<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<title>Cufflinks :: Center for Bioinformatics and Computational Biology</title>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8" />
<link rel="stylesheet" type="text/css" href="css/style.css" media="screen" />
</head>
<body>
<div id="wrap">
  <div id="top">
    <div class="lefts">
    <table width="100%" cellpadding="2">
      <tr><td>
        <a href="./index.html"><h1>Cufflinks</h1></a>
        <h2>Transcript assembly, differential expression, and differential regulation for RNA-Seq</h2>
      </td><td align="right" valign="middle">
        <h1><a href="http://www.cbcb.umd.edu/"><img border=0 src="images/cbcblogo.gif"></img></a>&nbsp;&nbsp;</h1>
      </td></tr>
    </table>
    </div>
  </div>
  <div id="main">
    <div id="rightside">
  		<h2>Site Map</h2>
		  <div class="box">
		    <ul>
		      <li><a href="index.html">Home</a></li>
		      <li><a href="tutorial.html">Getting started</a></li>
		      <li><a href="manual.html">Manual</a></li>
			  <li><a href="howitworks.html">How Cufflinks works</a></li>
		    </ul>
		  </div>
		
		  <h2><u>News and updates</u></h2>
		  <div class="box">
		    <ul>
		      <table width="100%">
				<tr><td>New releases and related tools will be announced through the <a href="https://lists.sourceforge.net/lists/listinfo/bowtie-bio-announce"><b>mailing list</b></a></td></tr>
			  </table>
		    </ul>
		  </div>
		
		  <a href="./downloads"><h2><u>Releases</u></h2></a>
		  <div class="box">
		    <ul>
		      <table width="100%">
				<tr><td>version 0.8.3 (BETA)</td> <td align="right">7/2/2010</td></tr>
				<tr>
		      		<td><a href="./downloads/cufflinks-0.8.3.tar.gz" onClick="javascript: pageTracker._trackPageview('/downloads/cufflinks_source'); ">&nbsp;&nbsp;&nbsp;Source code</a></td>
		      	</tr>
				<tr>
					<td><a href="./downloads/cufflinks-0.8.3.Linux_x86_64.tar.gz" onClick="javascript: pageTracker._trackPageview('/downloads/cufflinks'); ">&nbsp;&nbsp;&nbsp;Linux x86_64 binary</a></td>
				</tr>
				<tr>
					<td><a href="./downloads/cufflinks-0.8.3.OSX_x86_64.tar.gz" onClick="javascript: pageTracker._trackPageview('/downloads/cufflinks'); ">&nbsp;&nbsp;&nbsp;Mac OS X x86_64 binary</a></td>
				</tr>
			  </table>
		    </ul>
		  </div>

		  <h2>Related Tools</h2>
		  <div class="box">
		    <ul>
		      <li><a href="http://tophat.cbcb.umd.edu/">TopHat</a>: Alignment of short RNA-Seq reads</li>
		      <li><a href="http://bowtie.cbcb.umd.edu">Bowtie</a>: Ultrafast short read alignment</li>
		    </ul>
		  </div>

		  <h2>Publications</h2>
		  <div class="box">
		    <ul>
			  <li style="font-size: x-small; line-height: 130%"><p>Trapnell C, Williams BA, Pertea G, Mortazavi AM, Kwan G, van Baren MJ, Salzberg SL, Wold B, Pachter L.<b> <a href="http://dx.doi.org/10.1038/nbt.1621">Transcript assembly and quantification by RNA-Seq reveals unannotated transcripts and isoform switching during cell differentiation</b></a> <br/><i><a href="http://www.nature.com/nbt">Nature Biotechnology</a></i> doi:10.1038/nbt.1621</p><br/></li>
			  <li style="font-size: x-small; line-height: 130%"><p>Trapnell C, Pachter L, Salzberg SL. <a href="http://bioinformatics.oxfordjournals.org/cgi/content/abstract/btp120"><b>TopHat: discovering splice junctions with RNA-Seq</b></a>. <i><a href="http://bioinformatics.oxfordjournals.org">Bioinformatics</a></i>  doi:10.1093/bioinformatics/btp120</p><br/></li>
		      <li style="font-size: x-small; line-height: 130%"><p>Langmead B, Trapnell C, Pop M, Salzberg SL. <a href="http://genomebiology.com/2009/10/3/R25"><b>Ultrafast and memory-efficient alignment of short DNA sequences to the human genome</b></a>. <br/><i><a href="http://genomebiology.com">Genome Biology</a></i> <b>10</b>:R25.</li>
		    </ul>
		  </div>

		  <h2>Contributors</h2>
		  <div class="box">
		    <ul>
		      <li><a href="http://www.cs.umd.edu/~cole/">Cole Trapnell</a></li>
			  <li>Geo Pertea</li>
			  <li>Brian Williams</li>
			  <li><a href="http://wormlab.caltech.edu/members/">Ali Mortazavi</a></li>
			  <li>Gordon Kwan</a></li>  
			  <li>Jeltje van Baren</a></li>
		      <li><a href="http://www.cbcb.umd.edu/~salzberg/">Steven Salzberg</a></li>
		      <li><a href="http://biology.caltech.edu/Members/Wold">Barbara Wold</a></li>
			  <li><a href="http://www.math.berkeley.edu/~lpachter/">Lior Pachter</a></li>
		    </ul>
		  </div>

		  <h2>Links</h2>
		  <div class="box">
		    <ul>
		      <li><a href="http://bio.math.berkeley.edu/">Berkeley LMCB</a></li>
		      <li><a href="http://www.cbcb.umd.edu/">UMD CBCB</a></li>
			  <li><a href="http://woldlab.caltech.edu/">Wold Lab</a></li>
		    </ul>
		  </div>          

    </div> <!-- End of "rightside" -->
    <div id="leftside">
  	<table><tr><td cellpadding=7>
  	<h1> Getting started</h1><br/>
      <div id="toc">
  	    <ul>
  	    <li><a href="#inst">Install quick-start</a></li>
		<li><a href="#ref">Test the installation</a></li>
  	    <li><a href="#toph">Running Cufflinks</a></li>
  	    <li><a href="#exam">Examining your results</a></li>
  	    </ul><br/>
  	  </div>
	  	
	<br/>
  	<h2 id="inst">Install quick-start</h2><br/>
		<strong>Installing a pre-compiled binary release</strong><br/>
		<p>In order to make it easy to install Cufflinks, we provide a few binary
		packages to save users from occasionally frustrating process of building
		Cufflinks, which requires that you install the Boost libraries.  To use
		the binary packages, simply download the appropriate one for your machine,
		untar it, and place the <tt>cufflinks</tt> and <tt>cuffcompare</tt> binaries
		in your PATH environment variable.</p><br/>
		<strong>Building Cufflinks from source</strong><br/>
		<p>
		In order to build Cufflinks, you must have the 
		<a href="http://www.boost.org">Boost C++ libraries</a> (version 1.38 or higher) installed on 
		your system.  See below for instructions on installing Boost.
		</p>
		<br/>
		<h3 id="boost">Installing Boost</h3><br/>
		<ol>
		<li><a href="http://www.boost.org/users/download/">Download</a> Boost 
			and the <tt>bjam</tt> build engine.</li>
		<li>Unpack bjam and add it to your PATH.</li>
		<li>Unpack the Boost tarball and <tt>cd</tt> to the Boost source 
			directory.  This directory is called the BOOST_ROOT in some Boost
			installation instructions.</li>
		<li>Build Boost.  Note that you can specify where to put Boost with 
			the --prefix option. The default Boost installation directory is 
			<tt>/usr/local</tt>. Take note of the boost installation directory,
			because you will need to tell the Cufflinks installer where to find 
			Boost later on.  
			<p>If you are on Mac OS X, type (all on one line): <br/>
			<blockquote>bjam --prefix=&lt;YOUR_BOOST_INSTALL_DIRECTORY&gt; --toolset=darwin architecture=x86 address_model=32_64 link=static runtime-link=static --layout=versioned stage install</blockquote></li>
			<p>If you are on a 32-bit Linux system, type (all on one line): <br/>
				<blockquote>bjam --prefix=&lt;YOUR_BOOST_INSTALL_DIRECTORY&gt; --toolset=gcc architecture=x86 address_model=32 link=static runtime-link=static stage install</blockquote></li>
			<p>If you are on a 64-bit Linux system, type (all on one line): <br/>
				<blockquote>bjam --prefix=&lt;YOUR_BOOST_INSTALL_DIRECTORY&gt; --toolset=gcc architecture=x86 address_model=64 link=static runtime-link=static stage install</blockquote></li>
		</ol>
		</br>
			<h3 id="boost">Installing the SAM tools</h3><br/>
			<ol>
			<li><a href="http://samtools.sourceforge.net/">Download</a> the SAM tools</li>
			<li>Unpack the SAM tools tarball and <tt>cd</tt> to the SAM tools source 
				directory.</li>
			<li>Build the SAM tools by typing <tt>make</tt> at the command line.</li>
			<li>Choose a directory into which you wish to copy the SAM tools binary, the included library 
				<tt>libbam.a</tt>, and the library headers.  A common choice is <tt>/usr/local/</tt>.</lt>
			<li>Copy libbam.h to the <tt>lib/</tt> directory in the folder you've chosen above (e.g. <tt>/usr/local/lib/</tt>)</li>
			<li>Copy the headers (files ending in .h) to the <tt>include/</tt> directory you've chosen above (e.g. <tt>/usr/local/include/</tt>)
			<li>Copy the samtools binary to some directory in your PATH.</li>
			</ol>
		<br/>
		<h3 id="boost">Building Cufflinks</h3><br/>
		<ol>
			<li>Unpack the Cufflinks source tarball:
				<blockquote>tar zxvf cufflinks-0.7.0.tar.gz</blockquote></li><br/>
			<li>Change to the Cufflinks directory:
				<blockquote>cd cufflinks-0.7.0</blockquote></li><br/>
			<li>Configure Cufflinks.  If Boost is installed somewhere other than
				<tt>/usr/local</tt>, you will need to tell the installer where 
				to find it using the <tt>--with-boost</tt> option.  Specify where
				to install Cufflinks using the <tt>--prefix</tt> option.
				<blockquote>./configure --prefix=/path/to/cufflinks/install --with-boost=/path/to/boost</blockquote>
			    If you see any errors during configuration, verify that you 
				are using Boost version 1.38 or higher, and that the directory
				you specified via <tt>--with-boost</tt> contains the boost 
				header files and libraries.  See the Boost <a href="">Getting started</a> page for more details.  
				If you copied the SAM tools binaries to someplace other than <tt>/usr/local/</tt>, 
				you may need to supply the <tt>--with-bam</tt> configuration option.  
				<li>Finally, make and install Cufflinks.  
				<blockquote>make</blockquote>
				<blockquote>make install</blockquote>
				</blockquote>
			</li>
		</ol>
		
		
		<br/>
		<h2 id="test">Testing the installation</h2><br/>
		<ol> 
			<li>Download the <a href="downloads/test_data.sam">test data</a>
			<li>In the directory where you placed the test file, type:</br>
				<blockquote>cufflinks ./test_data.sam</blockquote><br/>
				You should see the following output:
				<pre>
Counting hits in map
	Total map density: 13725.000000
Processing bundle [ test_chromosome:52-550 ] with 99 non-redundant alignments
				</pre>
				
			<li>Verify that the file <tt>transcripts.gtf</tt> is in the current
				directory and looks like this (your file will have GTF attributes, omitted here for clarity)
				<pre>
test_chromosome Cufflinks       exon    53      250     1000    +       . 
test_chromosome Cufflinks       exon    351     400     1000    +       . 
test_chromosome Cufflinks       exon    501     550     1000    +       .
					
				</pre>
				</ol>
			</li>
      	
      	<h2 id="toph">Running Cufflinks</h2><br/>
		<p>
      	Please see the <a href="manual.html">user manual</a> for detailed 
		instructions on running Cufflinks.
		</p><br/>
    </div>
  </div>
  <div id="footer">
  	<table width="100%" cellspacing=15><tr><td>
    	This research was supported in part by NIH grants R01-LM06845 and R01-GM083873, NSF grant CCF-0347992 and the Miller Institute for Basic
			Research in Science at UC Berkeley.
    </td><td align=right>
    Administrator: <a href="mailto:cole@cs.umd.edu">Cole Trapnell</a>. Design by <a href="http://www.free-css-templates.com" title="Design by David Herreman">David Herreman</a>
    </td></tr></table>
  </div>
</div>

<!-- Google analytics code -->
<script type="text/javascript">
var gaJsHost = (("https:" == document.location.protocol) ? "https://ssl." : "http://www.");
document.write(unescape("%3Cscript src='" + gaJsHost + "google-analytics.com/ga.js' type='text/javascript'%3E%3C/script%3E"));
</script>
<script type="text/javascript">
try {
var pageTracker = _gat._getTracker("UA-6101038-2");
pageTracker._trackPageview();
} catch(err) {}</script>
<!-- End google analytics code -->

</body>
</html>
